File: SuperiorPackGroup/BLL/ShippingDetailsBLL.vb
Imports SuperiorPackGroup.SPGTableAdapters
Imports System.Text
Imports DevExpress.Xpo
Imports DXDAL.SPGData

<ComponentModel.DataObject()> _
Public Class ShippingDetailsBLL

    Private m_ShippingDetailsTableAdapter As ShippingDetailsTableAdapter = Nothing
    Public ReadOnly Property Adapter() As ShippingDetailsTableAdapter

        Get
            If m_ShippingDetailsTableAdapter Is Nothing Then
                m_ShippingDetailsTableAdapter = New ShippingDetailsTableAdapter()
                m_ShippingDetailsTableAdapter.Connection.ConnectionString = My.Settings.UserConnectionString
            End If
            Return m_ShippingDetailsTableAdapter
        End Get

    End Property

    Private Sub UpdateAuditTrail(ByVal ModifiedRecord As SPG.ShippingDetailsRow, ByVal originalRecord As Object())

        Dim builder As New StringBuilder(String.Empty)
        Dim recordFields As Integer = (ModifiedRecord.ItemArray.Length - 1)
        Dim i As Integer = 0
        Do While (i <= recordFields)
            Try
                If IsDBNull(originalRecord(i)) Then
                    If Not IsDBNull(ModifiedRecord.Item(i)) Then
                        builder.Append(String.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns.Item(i).ColumnName, "NULL", ModifiedRecord.Item(i)))
                    End If
                ElseIf IsDBNull(ModifiedRecord.Item(i)) Then
                    If Not IsDBNull(originalRecord(i)) Then
                        builder.Append(String.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns.Item(i).ColumnName, originalRecord(i), "NULL"))
                    End If
                ElseIf ModifiedRecord.Item(i) IsNot originalRecord(i) Then
                    builder.Append(String.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns.Item(i).ColumnName, originalRecord(i), ModifiedRecord.Item(i)))
                End If
            Catch
            End Try
            i += 1
        Loop
        If (builder.Length > 2) Then
            builder.Length = (builder.Length - 2)
            AuditTrailBLL.AddTrailEntry(My.Settings.UserName, CType(ModifiedRecord.Item(0), Integer), "ShippingDetails", builder.ToString)
        End If

    End Sub

    <ComponentModel.DataObjectMethod(ComponentModel.DataObjectMethodType.Select, False)> _
    Public Function GetShippingDetailsByItemID(ByVal itemId As Integer) As SPG.ShippingDetailsDataTable

        Return Adapter.GetShippingDetailsByItemID(itemId)

    End Function

    <ComponentModel.DataObjectMethod(ComponentModel.DataObjectMethodType.Select, False)> _
    Public Function GetDetailsByShippingID(ByVal shippingID As Integer) As SPG.ShippingDetailsDataTable

        Return Adapter.GetDetailsByShippingID(shippingID)

    End Function

    <ComponentModel.DataObjectMethod(ComponentModel.DataObjectMethodType.Update, True)> _
    Public Function UpdateShippingDetails(ByVal session As Session, ByVal detailID As Integer?, ByVal shippingID As Integer, ByVal itemID As Integer?, ByVal lot As String, ByVal quantity As Integer?, _
                                          ByVal pallets As Single?, ByVal note As String, ByVal expirationDate As Date?, ByVal fullLPNNumber As String) As Boolean

        If Not itemID.HasValue Then
            Throw New ApplicationException("You must provide shipping item.")
        End If

        If Not quantity.HasValue Then
            Throw New ApplicationException("You must provide the Quantity shipped.")
        End If

        Dim details As SPG.ShippingDetailsDataTable = Adapter.GetDetailsByDetailID(detailID.Value)

        If details.Count = 0 Then
            'It is a new Detail
            Return Me.InsertDetails(session, shippingID, itemID, lot, quantity, pallets, note, expirationDate, fullLPNNumber)
        End If

        Dim shippingDetail As SPG.ShippingDetailsRow = details(0)
        Dim originalItem, originalQuantity, newQuantity As Integer
        Dim itemChanged As Boolean = False

        Dim originalRecord As Object() = Nothing
        originalRecord = shippingDetail.ItemArray

        Dim originalDetail As SPG.ShippingDetailsRow = details.NewShippingDetailsRow()
        originalDetail.ItemArray = CType(originalRecord.Clone(), Object())

        originalItem = shippingDetail.ShipDetItemID
        If originalItem <> itemID.Value Then
            itemChanged = True
        End If
        shippingDetail.ShipDetItemID = itemID.Value
        If String.IsNullOrEmpty(lot) Then
            shippingDetail.SetShipDetLotNull()
        Else
            shippingDetail.ShipDetLot = lot
        End If
        originalQuantity = shippingDetail.ShipDetDetQty
        newQuantity = quantity.Value - shippingDetail.ShipDetDetQty
        shippingDetail.ShipDetDetQty = quantity.Value
        shippingDetail.SetsngPackagesNull()
        If Not pallets.HasValue Then
            shippingDetail.SetsngPalletsNull()
        Else
            shippingDetail.sngPallets = pallets.Value
        End If
        If String.IsNullOrEmpty(note) Then
            shippingDetail.SetNoteNull()
        Else
            shippingDetail.Note = note
        End If
        If expirationDate.HasValue Then
            shippingDetail.ExpirationDate = expirationDate.Value
        Else
            shippingDetail.SetExpirationDateNull()
        End If
        If String.IsNullOrEmpty(fullLPNNumber) Then
            shippingDetail.SetFullLPNNumberNull()
        Else
            shippingDetail.FullLPNNumber = fullLPNNumber
        End If

        If Not IsNothing(originalRecord) Then
            Me.UpdateAuditTrail(shippingDetail, originalRecord)
        End If

        Dim rowsAffected As Integer = Adapter.Update(shippingDetail)

        If rowsAffected = 1 Then
            Dim items As ItemsBLL = New ItemsBLL
            Dim locationID As Integer = session.GetObjectByKey(Of Shipping)(shippingID).ShippingLocation.Oid
            'If itemChanged = True Then
            If originalDetail.FullLPNNumber.StartsWith(CustomersBLL.GetLPNPrefix(7)) Then
                items.UpdateStock(session, originalItem, originalQuantity, False, locationID, originalDetail.ShipDetLot, Integer.Parse(originalDetail.FullLPNNumber.Replace(CustomersBLL.GetLPNPrefix(7), "")))
            Else
                items.UpdateStock(session, originalItem, originalQuantity, False, locationID)
                End If
                If fullLPNNumber.StartsWith(CustomersBLL.GetLPNPrefix(7)) Then
                    items.UpdateStock(session, itemID.Value, quantity.Value * -1, False, locationID, lot, Integer.Parse(fullLPNNumber.Replace(CustomersBLL.GetLPNPrefix(7), "")))
                Else
                    items.UpdateStock(session, itemID.Value, quantity.Value * -1, False, locationID)
                End If
                'Else
                '    If newQuantity <> 0 Then
                '        items.UpdateStock(session, itemID.Value, newQuantity * -1, False, locationID)
                '    End If
                'End If
            End If

        Return rowsAffected = 1

    End Function

    <ComponentModel.DataObjectMethod(ComponentModel.DataObjectMethodType.Insert, True)> _
    Public Function InsertDetails(ByVal session As Session, ByVal shippingID As Integer, ByVal itemID As Integer?, ByVal lot As String, ByVal quantity As Integer?, ByVal pallets As Single?, _
                                  ByVal note As String, ByVal expirationDate As Date?, ByVal fullLPNNumber As String) As Boolean

        Dim details As SPG.ShippingDetailsDataTable = New SPG.ShippingDetailsDataTable
        Dim shippingDetail As SPG.ShippingDetailsRow = details.NewShippingDetailsRow()

        shippingDetail.ShipDetMainID = shippingID
        shippingDetail.ShipDetItemID = itemID.Value
        If String.IsNullOrEmpty(lot) Then
            shippingDetail.SetShipDetLotNull()
        Else
            shippingDetail.ShipDetLot = lot
        End If
        shippingDetail.ShipDetDetQty = quantity.Value
        shippingDetail.SetsngPackagesNull()
        If Not pallets.HasValue Then
            shippingDetail.SetsngPalletsNull()
        Else
            shippingDetail.sngPallets = pallets.Value
        End If
        If String.IsNullOrEmpty(note) Then
            shippingDetail.SetNoteNull()
        Else
            shippingDetail.Note = note
        End If
        If expirationDate.HasValue Then
            shippingDetail.ExpirationDate = expirationDate.Value
        Else
            shippingDetail.SetExpirationDateNull()
        End If
        If String.IsNullOrEmpty(fullLPNNumber) Then
            shippingDetail.SetFullLPNNumberNull()
        Else
            shippingDetail.FullLPNNumber = fullLPNNumber
        End If

        details.AddShippingDetailsRow(shippingDetail)
        Dim rowsAffected As Integer = Adapter.Update(details)

        If rowsAffected = 1 Then
            Dim items As ItemsBLL = New ItemsBLL
            If fullLPNNumber.StartsWith(CustomersBLL.GetLPNPrefix(7)) Then
                items.UpdateStock(session, itemID.Value, quantity.Value * -1, False, session.GetObjectByKey(Of Shipping)(shippingID).ShippingLocation.Oid, lot, Integer.Parse(fullLPNNumber.Replace(CustomersBLL.GetLPNPrefix(7), "")))
            Else
                items.UpdateStock(session, itemID.Value, quantity.Value * -1, False, session.GetObjectByKey(Of Shipping)(shippingID).ShippingLocation.Oid)
            End If
        End If

        Return rowsAffected = 1

    End Function

    <ComponentModel.DataObjectMethod(ComponentModel.DataObjectMethodType.Delete, True)> _
    Public Function DeleteShippingDetail(ByVal session As Session, ByVal detailID As Integer) As Boolean

        Dim details As SPG.ShippingDetailsDataTable = Adapter.GetDetailsByDetailID(detailID)
        Dim rowsAffected As Integer = 0
        Dim itemID, quantity, locationID As Integer
        Dim lot, fullLPNNumber As String

        If details.Count = 1 Then
            Dim shippingDetail As SPG.ShippingDetailsRow = CType(details.Rows(0), SPG.ShippingDetailsRow)
            itemID = shippingDetail.ShipDetItemID
            quantity = shippingDetail.ShipDetDetQty
            lot = shippingDetail.ShipDetLot
            fullLPNNumber = shippingDetail.FullLPNNumber
            locationID = session.GetObjectByKey(Of Shipping)(shippingDetail.ShipDetMainID).ShippingLocation.Oid
            rowsAffected = Adapter.Delete(detailID, shippingDetail.ts)
        End If

        If rowsAffected = 1 Then
            Dim items As ItemsBLL = New ItemsBLL
            If fullLPNNumber.StartsWith(CustomersBLL.GetLPNPrefix(7)) Then
                items.UpdateStock(session, itemID, quantity, False, locationID, lot, Integer.Parse(fullLPNNumber.Replace(CustomersBLL.GetLPNPrefix(7), "")))
            Else
                items.UpdateStock(session, itemID, quantity, False, locationID)
            End If
        End If

        'Return true if precisely one row was deleted, otherwise return false.
        Return rowsAffected = 1

    End Function

End Class

File: SuperiorPackGroup/BLL/ShippingDetailsBLL.cs
//INSTANT C# NOTE: Formerly VB project-level imports:
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;

using SuperiorPackGroup.SPGTableAdapters;
using System.Text;
using DevExpress.Xpo;
using DXDAL.SPGData;
using DevExpress.Data.Filtering;
using System.Linq;

namespace SuperiorPackGroup
{
	[System.ComponentModel.DataObject()]
	public class ShippingDetailsBLL
	{

		private ShippingDetailsTableAdapter m_ShippingDetailsTableAdapter = null;
		public ShippingDetailsTableAdapter Adapter
		{

			get
			{
				if (m_ShippingDetailsTableAdapter == null)
				{
					m_ShippingDetailsTableAdapter = new ShippingDetailsTableAdapter();
					m_ShippingDetailsTableAdapter.Connection.ConnectionString = Properties.Settings.Default.UserConnectionString;
				}
				return m_ShippingDetailsTableAdapter;
			}

		}

		private void UpdateAuditTrail(SPG.ShippingDetailsRow ModifiedRecord, object[] originalRecord)
		{

			StringBuilder builder = new StringBuilder(string.Empty);
			int recordFields = (ModifiedRecord.ItemArray.Length - 1);
			int i = 0;
			while (i <= recordFields)
			{
				try
				{
					if (Convert.IsDBNull(originalRecord[i]))
					{
						if (!Convert.IsDBNull(ModifiedRecord[i]))
						{
							builder.Append(string.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns[i].ColumnName, "NULL", ModifiedRecord[i]));
						}
					}
					else if (Convert.IsDBNull(ModifiedRecord[i]))
					{
						if (!Convert.IsDBNull(originalRecord[i]))
						{
							builder.Append(string.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns[i].ColumnName, originalRecord[i], "NULL"));
						}
					}
					else if (ModifiedRecord[i] != originalRecord[i])
					{
						builder.Append(string.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns[i].ColumnName, originalRecord[i], ModifiedRecord[i]));
					}
				}
				catch
				{
				}
				i += 1;
			}
			if (builder.Length > 2)
			{
				builder.Length = (builder.Length - 2);
				AuditTrailBLL.AddTrailEntry(Properties.Settings.Default.UserName, Convert.ToInt32(ModifiedRecord[0]), "ShippingDetails", builder.ToString());
			}

		}

		[System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Select, false)]
		public SPG.ShippingDetailsDataTable GetDetailsByShippingID(int shippingID)
		{

			return Adapter.GetDetailsByShippingID(shippingID);

		}

		[System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Update, true)]
		public bool UpdateShippingDetails(Session session, int? detailID, int shippingID, int? itemID, string lot, int? quantity, float? pallets, string note, DateTime? expirationDate, string fullLPNNumber)
		{

			if (!itemID.HasValue)
			{
				throw new ApplicationException("You must provide shipping item.");
			}

			if (!quantity.HasValue)
			{
				throw new ApplicationException("You must provide the Quantity shipped.");
			}

			SPG.ShippingDetailsDataTable details = Adapter.GetDetailsByDetailID(detailID.Value);

			if (details.Count() == 0)
			{
				//It is a new Detail
				return this.InsertDetails(session, shippingID, itemID, lot, quantity, pallets, note, expirationDate, fullLPNNumber);
			}

			SPG.ShippingDetailsRow shippingDetail = details[0];
			int originalItem = 0;
			int originalQuantity = 0;
			int newQuantity = 0;
			bool itemChanged = false;

			object[] originalRecord = shippingDetail.ItemArray;

			SPG.ShippingDetailsRow originalDetail = details.NewShippingDetailsRow();
			originalDetail.ItemArray = (object[])originalRecord.Clone();

			originalItem = shippingDetail.ShipDetItemID;
			if (itemID.HasValue && originalItem != itemID.Value)
			{
				itemChanged = true;
			}
			shippingDetail.ShipDetItemID = itemID.Value;
			if (string.IsNullOrEmpty(lot))
			{
				shippingDetail.SetShipDetLotNull();
			}
			else
			{
				shippingDetail.ShipDetLot = lot;
			}
			originalQuantity = shippingDetail.ShipDetDetQty;
			newQuantity = quantity.Value - shippingDetail.ShipDetDetQty;
			shippingDetail.ShipDetDetQty = quantity.Value;
			shippingDetail.SetsngPackagesNull();
			if (!pallets.HasValue)
			{
				shippingDetail.SetsngPalletsNull();
			}
			else
			{
				shippingDetail.sngPallets = pallets.Value;
			}
			if (string.IsNullOrEmpty(note))
			{
				shippingDetail.SetNoteNull();
			}
			else
			{
				shippingDetail.Note = note;
			}
			if (expirationDate.HasValue)
			{
				shippingDetail.ExpirationDate = expirationDate.Value;
			}
			else
			{
				shippingDetail.SetExpirationDateNull();
			}
			if (string.IsNullOrEmpty(fullLPNNumber))
			{
				shippingDetail.SetFullLPNNumberNull();
			}
			else
			{
				shippingDetail.FullLPNNumber = fullLPNNumber;
			}

			if (!(originalRecord == null))
			{
				this.UpdateAuditTrail(shippingDetail, originalRecord);
			}

			int rowsAffected = Adapter.Update(shippingDetail);

			if (rowsAffected == 1)
			{
				ItemsBLL items = new ItemsBLL();
				int locationID = session.GetObjectByKey<Shipping>(shippingID).ShippingLocation.Oid;
				//If itemChanged = True Then
				if (originalDetail.FullLPNNumber.StartsWith(CustomersBLL.GetLPNPrefix(7)))
				{
					items.UpdateStock(session, originalItem, originalQuantity, false, locationID, originalDetail.ShipDetLot, int.Parse(originalDetail.FullLPNNumber.Replace(CustomersBLL.GetLPNPrefix(7), "")));
				}
				else
				{
					items.UpdateStock(session, originalItem, originalQuantity, false, locationID);
				}
				if (fullLPNNumber.StartsWith(CustomersBLL.GetLPNPrefix(7)))
				{
					items.UpdateStock(session, itemID.Value, quantity.Value * -1, false, locationID, lot, int.Parse(fullLPNNumber.Replace(CustomersBLL.GetLPNPrefix(7), "")));
				}
				else
				{
					items.UpdateStock(session, itemID.Value, quantity.Value * -1, false, locationID);
				}
				//Else
				//    If newQuantity <> 0 Then
				//        items.UpdateStock(session, itemID.Value, newQuantity * -1, False, locationID)
				//    End If
				//End If
			}

			return rowsAffected == 1;

		}

		[System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Insert, true)]
		public bool InsertDetails(Session session, int shippingID, int? itemID, string lot, int? quantity, float? pallets, string note, DateTime? expirationDate, string fullLPNNumber)
		{

			SPG.ShippingDetailsDataTable details = new SPG.ShippingDetailsDataTable();
			SPG.ShippingDetailsRow shippingDetail = details.NewShippingDetailsRow();

			shippingDetail.ShipDetMainID = shippingID;
			shippingDetail.ShipDetItemID = itemID.Value;
			if (string.IsNullOrEmpty(lot))
			{
				shippingDetail.SetShipDetLotNull();
			}
			else
			{
				shippingDetail.ShipDetLot = lot;
			}
			shippingDetail.ShipDetDetQty = quantity.Value;
			shippingDetail.SetsngPackagesNull();
			if (!pallets.HasValue)
			{
				shippingDetail.SetsngPalletsNull();
			}
			else
			{
				shippingDetail.sngPallets = pallets.Value;
			}
			if (string.IsNullOrEmpty(note))
			{
				shippingDetail.SetNoteNull();
			}
			else
			{
				shippingDetail.Note = note;
			}
			if (expirationDate.HasValue)
			{
				shippingDetail.ExpirationDate = expirationDate.Value;
			}
			else
			{
				shippingDetail.SetExpirationDateNull();
			}
			if (string.IsNullOrEmpty(fullLPNNumber))
			{
				shippingDetail.SetFullLPNNumberNull();
			}
			else
			{
				shippingDetail.FullLPNNumber = fullLPNNumber;
			}

			details.AddShippingDetailsRow(shippingDetail);
			int rowsAffected = Adapter.Update(details);

			if (rowsAffected == 1)
			{
				ItemsBLL items = new ItemsBLL();
				if (fullLPNNumber.StartsWith(CustomersBLL.GetLPNPrefix(7)))
				{
					items.UpdateStock(session, itemID.Value, quantity.Value * -1, false, session.GetObjectByKey<Shipping>(shippingID).ShippingLocation.Oid, lot, int.Parse(fullLPNNumber.Replace(CustomersBLL.GetLPNPrefix(7), "")));
				}
				else
				{
					items.UpdateStock(session, itemID.Value, quantity.Value * -1, false, session.GetObjectByKey<Shipping>(shippingID).ShippingLocation.Oid);
				}
			}

			return rowsAffected == 1;

		}

		[System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Delete, true)]
		public bool DeleteShippingDetail(Session session, int detailID)
		{

			SPG.ShippingDetailsDataTable details = Adapter.GetDetailsByDetailID(detailID);
			int rowsAffected = 0;
			int itemID = 0;
			int quantity = 0;
			int locationID = 0;
			string lot = null;
			string fullLPNNumber = null;

			if (details.Count() == 1)
			{
				SPG.ShippingDetailsRow shippingDetail = (SPG.ShippingDetailsRow)details.Rows[0];
				itemID = shippingDetail.ShipDetItemID;
				quantity = shippingDetail.ShipDetDetQty;
				lot = shippingDetail.ShipDetLot;
				fullLPNNumber = shippingDetail.FullLPNNumber;
				locationID = session.GetObjectByKey<Shipping>(shippingDetail.ShipDetMainID).ShippingLocation.Oid;
				rowsAffected = Adapter.Delete(detailID, shippingDetail.ts);
			}

			if (rowsAffected == 1)
			{
				ItemsBLL items = new ItemsBLL();
				if (fullLPNNumber.StartsWith(CustomersBLL.GetLPNPrefix(7)))
				{
					items.UpdateStock(session, itemID, quantity, false, locationID, lot, int.Parse(fullLPNNumber.Replace(CustomersBLL.GetLPNPrefix(7), "")));
				}
				else
				{
					items.UpdateStock(session, itemID, quantity, false, locationID);
				}
			}

			//Return true if precisely one row was deleted, otherwise return false.
			return rowsAffected == 1;

		}
		public static XPView GetShippingDetailsByItemID(int itemID)
		{

			XPView ShippingXPView = new XPView(Session.DefaultSession, typeof(ShipDet));

			ShippingXPView.Properties.AddRange(new ViewProperty[]
			{
				new ViewProperty(ShipDet.Fields.ShipDetDetID.PropertyName, SortDirection.None, ShipDet.Fields.ShipDetDetID, false, true),
				new ViewProperty(ShipDet.Fields.ShipDetMainID.ShipMainID.PropertyName, SortDirection.Ascending, ShipDet.Fields.ShipDetMainID.ShipMainID, false, true),
				new ViewProperty(ShipDet.Fields.ShipDetLot.PropertyName, SortDirection.None, ShipDet.Fields.ShipDetLot, false, true),
				new ViewProperty(ShipDet.Fields.ExpirationDate.PropertyName, SortDirection.None, ShipDet.Fields.ExpirationDate, false, true),
				new ViewProperty(ShipDet.Fields.ShipDetDetQty.PropertyName, SortDirection.None, ShipDet.Fields.ShipDetDetQty, false, true),
				new ViewProperty("ShipMainBOL", SortDirection.None, ShipDet.Fields.ShipDetMainID.ShipMainBOL, false, true),
				new ViewProperty("ShipMainDate", SortDirection.None, ShipDet.Fields.ShipDetMainID.ShipMainDate, false, true)
			});

			ShippingXPView.Criteria = CriteriaOperator.And(new BinaryOperator(ShipDet.Fields.ShipDetItemID.ItemType.PropertyName, "FG", BinaryOperatorType.Equal), new BinaryOperator(ShipDet.Fields.ShipDetItemID.ItemID.PropertyName, itemID, BinaryOperatorType.Equal));
			return ShippingXPView;

		}
	}

}